A32-xds-circuit-breaking.md

```diff
@@ -0,0 +1,157 @@
+at any given time, in `Thresholds` is considered. All the others are either 
+inapplicable to gRPC or the corresponding features are missing for now. The
+design and implementation section of this proposal mainly talks about the
+enforcement of the limit for maximum number of outstanding requests sent to
+an upstream cluster. See [Other configurations considered](#other-configurations-considered) 
+for rational.
+
+To match Envoy's implementation, circuit breaking limits will always be 
+enforced with certain defaults, even if the gRPC client receives no applicable 
+circuit breakers. Users can effectively turn circuit breaking off by setting
+thresholds to very high values, for example, to 
+`std::numeric_limits<uint32_t>::max()`.
+
+## Detailed design
+
+### Maintaining thresholds
+
+Each CDS LB policy maintains a threshold for the maximum number of outstanding 
+requests to hosts in the cluster that this policy is load balancing for. Make
+sure this threshold can be updated when receiving an update from the
+XdsClient cluster watch interface and the updated value can be dynamically 
+picked up by request limiting logic. By default, this value is set to 1024.
+
+### Limiting outstanding requests
+
+Each CDS LB policy uses a counter to count the number of requests sent through
+the _subchannels_ picked by its _Picker_, which is the output of the CDS LB
+policy. The counter has the same lifecycle as the CDS LB policy. In the _Pick_
+operation, the outputted Picker instance should includes the logic of checking 
+if the counter's value reaches the threshold before delegating the pick to the 
+Picker of downstream LB policies. If the threshold is reached, it returns
+a result that leads to the request being dropped with `UNAVAIABLE` status and
+never gets retried. Also this drop is recorded and aggregated to the 
+`total_dropped_requests` in the cluster-level stats to be reported in
+load reporting via LRS.
+
+The counter is incremented when creating the stream for the request and 
+decremented the stream is closed. A general implementation can increment the
+counter at the _Pick_ operation of the outputted Picker for the case the 
+threshold is not reached. The callback of stream closed takes the 
+responsibility of decrementing the counter. Each language may have its own
+mechanism for tracking stream created/closed (e.g., in Java it has 
+`ClientStreamTracer`) and implementations may choose to increment/decrement the
+counter at the most suitable places. Special considerations need to be taken
+for:
+- There can be a race between checking the counter's value in making a pick
+decision and the counter's value being mutated.
+- An increment should always have a corresponding decrement. A common mistake
+can be the counter is incremented in the `Pick` operation while the pick result
+is discarded later, which leads to the value being leaked.
+- The counter should only be incremented if the pick result indeed leads to a
```

Comment:
This implies that the CDS picker should call the EDS picker first and then decide whether to discard the result.  That seems a little inefficient; we shouldn't need to go through the trouble of calling the EDS picker if we're just going to discard its result anyway.

I think that if the current number of in-flight calls is already at the limit, then we can immediately drop any new picks without calling the EDS policy.  There's no need to call the EDS policy if we're going to drop the call anyway.

Reply:
Yes you are right, but that's for the drop due to hitting the circuit breaker limit. The case I am talking about here is the EDS picker can also return a result that is a drop (a.k.a., contains no Subchannel, or even an error result). In that case, the counter should not be incremented as this will never leads to a RPC being sent. In Java, we probably do not have this concern as we manipulate the counter in stream tracer (a.k.a., a callback object that listens on stream events such as stream created/closed). But in C-core and Go, the counter is incremented in the Pick operation, so implementations should make sure the pick does lead to an RPC being sent before incrementing the counter

Reply:
Ah, okay -- sorry, I misunderstood what this said when I first read it.

It's not clear to me why this needs to be pointed out, because if the CDS policy itself is not deciding to drop the request, it would certainly not update its counter.

In any case, as I mentioned elsewhere, I think we can omit some of the detailed implementation concerns from this doc, since they're mainly of concern to implementors.

Reply:
Deleted details that are not directly related to CDS policy and its picker.

Reply:
I don't actually see anything deleted here.  There's still a lot of text related to implementation details.

I don't necessarily mind keeping these details here if you want, but how about moving them into their own subsection, with a heading like "Implementation Considerations"?

Reply:
Sure, I eliminated the whole paragraph. It is implementation specific, readers with only high-level background may not be able to get it. Those can (have been) be discussed in our own design doc for implementors.